pkg/segment/query/querystatus.go

```diff
@@ -176,18 +206,70 @@ func StartQuery(qid uint64, async bool, cleanupCallback func()) (*RunningQuerySt
 		return nil, fmt.Errorf("qid has already been started")
 	}
 
-	if len(allRunningQueries) >= MAX_RUNNING_QUERIES {
-		return nil, putils.TeeErrorf("StartQuery: qid=%v cannot be started, Max number of running queries reached", qid)
+	stateChan := make(chan *QueryStateChanData, numStates)
+
+	runningState := &RunningQueryState{
+		startTime:         time.Now(),
+		StateChan:         stateChan,
+		cleanupCallback:   cleanupCallback,
+		rqsLock:           &sync.Mutex{},
+		isAsync:           async,
+		timeoutCancelFunc: nil,
+	}
+
+	waitingQueriesLock.Lock()
+	defer waitingQueriesLock.Unlock()
+	if len(waitingQueries) >= MAX_WAITING_QUERIES {
+		return nil, putils.TeeErrorf("StartQuery: qid=%v cannot be started, Max number of waiting queries reached", qid)
 	}
+	waitingQueries = append(waitingQueries, &WaitStateData{qid, runningState})
+
+	return runningState, nil
+}
 
-	var stateChan chan *QueryStateChanData
-	if async {
-		stateChan = make(chan *QueryStateChanData, numStates)
-		stateChan <- &QueryStateChanData{StateName: RUNNING}
+// Removes reference to qid. If qid does not exist this is a noop
+func DeleteQuery(qid uint64) {
+	LogGlobalSearchErrors(qid)
+	arqMapLock.Lock()
+	rQuery, ok := allRunningQueries[qid]
+	if ok {
+		delete(allRunningQueries, qid)
 	}
+	arqMapLock.Unlock()
 
+	if ok && !rQuery.isCancelled {
+		rQuery.timeoutCancelFunc()
+
+		if rQuery.cleanupCallback != nil {
+			rQuery.cleanupCallback()
+		}
+	}
+}
+
+func canRunQuery() bool {
+	activeQueries := uint64(GetActiveQueryCount())
+	return activeQueries < MAX_RUNNING_QUERIES
+}
+
+func PullQueriesToRun() {
+	for {
+		if canRunQuery() {
+			waitingQueriesLock.Lock()
+			if len(waitingQueries) == 0 {
+				waitingQueriesLock.Unlock()
+				continue
```

Comment:
when this go routine is lanuched, if there are no queries in the system, then there is spiining loop happening here, and 1 cpu core will be at 100% even though no queries are running at all.

Reply:
we should add a sleep even here before doing continue. 
@AndrewHess thoughts ?

Reply:
addressed in the latest commit.